### ISSUE TITLE
Lines sent to client are not truncated to RFC spec

### ISSUE THREAD

**Issue by richardg867** [inspircd/inspircd#1101]
InspIRCd will truncate any lines sent to a client to 512 bytes plus CRLF, violating RFC 1459:

> IRC messages are always lines of characters terminated with a CR-LF
> (Carriage Return - Line Feed) pair, and these messages **shall not
> exceed 512 characters in length, counting all characters including
> the trailing CR-LF.** Thus, there are 510 characters maximum allowed
> for the command and its parameters.  There is no provision for
> continuation message lines.

This incorrect truncation causes issues with some legacy clients upon receiving a 512 byte long line, which, after adding the CRLF, is 514 bytes long. InspIRCd should truncate lines to 510 bytes plus CRLF instead in order to be RFC compliant.

**Comment by rburchell** [comment]
Do you have a protocol debug of this?

The reason I ask is that it should be doing the right thing, at a quick look:
src/users.cpp, LocalUser::Write:

	if (text.length() > ServerInstance->Config->Limits.MaxLine - 2)
	{
		// this should happen rarely or never. Crop the string at 512 and try again.
		std::string try_again(text, 0, ServerInstance->Config->Limits.MaxLine - 2);
		Write(try_again);
		return;
	}

Where MaxLine is defined as:

src/configreader.cpp:   , MaxLine(tag->getInt("maxline", 512))

So, unless you have a custom maxline setting, or something else is doing the writing by hand, I think this should be working...

**Comment by rburchell** [comment]
On a tangent, I'm also curious what sort of issues you've seen in clients.

**Comment by richardg867** [comment]
I tested the issue by sending a message (on a InspIRCd-2.0 server with no maxline defined) with the character 'A' repeated more than 512 times, and listening with another client. The other client received:

```:t!t@RoL-ub2.vfr.158.179.IP PRIVMSG #test :AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\r\n```

...which, considering the trailing \r\n as a CRLF, is 514 bytes long.

What caused me to investigate the issue is a buffer overflow in the old Microsoft Chat client caused by such long messages.

**Comment by Sheogorath-SI** [comment]
You're right.

in Insp20 Branch in file win/inspircd_config.h.cmake MAXBUF is definied as 514 and used in LocalUser::Write:

```
if (text.length() > MAXBUF - 2)
	{
		// this should happen rarely or never. Crop the string at 512 and try again.
		std::string try_again = text.substr(0, MAXBUF - 2);
		Write(try_again);
		return;
	}
``` 

Should be changed

**Comment by rburchell** [comment]
2.0 (and Windows) are indeed useful pieces of information. Good spot, @Sheogorath-SI :)

**Comment by Sheogorath-SI** [comment]
I can "fix" it by lower the MAXBUF and give it a try on my network. if there is no other solution.

**Comment by Adam-** [comment]
This isn't specific to just Windows; configure also sets it to 514.

**Comment by SaberUK** [comment]
I fixed this in master already with the move to `<limits:maxline>` but unfortunately can't fix it in 2.0 for compatibility reasons.

**Comment by rburchell** [comment]
@Adam- I see 512 in configure line 275. Is there somewhere else I should be looking?

**Comment by Adam-** [comment]
@rburchell https://github.com/inspircd/inspircd/blob/insp20/configure#L857

**Comment by attilamolnar** [comment]
It's ok to fix this in 2.0 as this is clearly breaking the RFC as shown by @richardg867.
@Adam-  pushed a commit that fixes this, can you please take a look?

https://github.com/Adam-/inspircd/commit/fd17c9a964597747265cb0fa750aad6e13f3774f

**Comment by rburchell** [comment]
Change looks fine to me, although it's hard to be absolutely certain without vetting all the users of MAXBUF..

**Comment by attilamolnar** [comment]
Yes, I'm contemplating changing the relevant truncation code in users.cpp only.

**Comment by rburchell** [comment]
That would strike me as a prudent suggestion, although if e.g. RPL_NAMREPLY is using MAXBUF, you might lose some off the end..

**Comment by attilamolnar** [comment]
The proposed fix needs more testing due to MAXBUF being used in many places throughout the code.

**Comment by SaberUK** [comment]
This has been fixed in the master branch. Unfortunately it is not possible to fix in 2.0 due to how many things it might break.
